### PR TITLE
Structural/add custom scipy solver

### PR DESCRIPTION
**Pull request by manuelmessmer**
Provides stiffness and mass matrix as scipy matrices. 
This pull request shall resolve the issue #6111 opened by @RiccardoRossi . The implementation of the solver is pretty much what Riccardo suggested as a prototype.

### CODE REVIEW

File: kratos/python/add_strategies_to_python.cpp [hunk 1]
```diff
@@ -387,6 +387,7 @@ namespace Kratos
                 .def("BuildAndSolve", &BuilderAndSolverType::BuildAndSolve)
                 .def("BuildRHSAndSolve", &BuilderAndSolverType::BuildRHSAndSolve)
                 .def("ApplyDirichletConditions", &BuilderAndSolverType::ApplyDirichletConditions)
+                .def("ApplyConstraints", &BuilderAndSolverType::ApplyConstraints)
```
Comment (loumalouomega) [inline kratos/python/add_strategies_to_python.cpp]: This changes should be in a independent PR..m the rest is OK

**Review by loumalouomega — DISMISSED** [review]
Ok from my side

**Review by philbucher — CHANGES_REQUESTED** [review]
Cool!
Blocking as I need some days to review

**Comment by qaumann** [comment]
Thanks for the changes, I think this is very useful.
I am just generally wondering about the building process. Is the workaround using BUILD_LEVEL the way to go for this kind of problem or should a builder and solver provide the methods to build the different system matrices (e.g. BuildMassMatrix, BuildStiffnessMatrix, etc.)?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    
+    Derived class must override the function SolveSolutionStep. In there the Mass and Stiffness matrices
+    can be obtained as scipy matrices.
+
+    See structural_mechanics_solver.py for more information.
+    """
+    def __init__(self, main_model_part, custom_settings):
+        # Construct the base solver.
+        super(CustomScipySolver, self).__init__(main_model_part, custom_settings)
+        KratosMultiphysics.Logger.PrintInfo("::[CustomScipySolver]:: ", "Construction finished")
```
Comment (armingeiser) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: i would move this message to the bottom of the constructor

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            solution_scheme = StructuralMechanicsApplication.EigensolverDynamicScheme()
+        else: # here e.g. a stability scheme could be added
+            err_msg =  "The requested scheme type \"" + scheme_type + "\" is not available!\n"
+            err_msg += "Available options are: \"dynamic\""
+            raise Exception(err_msg)
+
+        return solution_scheme
+
+    def _create_linear_solver(self):
+        ''' Linear solver will not be used. But eventually the solution stratgey calls the solver's clear function.
```
Comment (armingeiser) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]:
```suggestion
        ''' Linear solver will not be used. But eventually the solution strategy calls the solver's clear function. 
```

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        # Apply boundary conditions
+        builder_and_solver.ApplyDirichletConditions(scheme, self.GetComputingModelPart(), aux, xD, b)
+        # Convert stiffness matrix to scipy
+        K = KratosMultiphysics.scipy_conversion_tools.to_csr(aux)
+
+        return K
+        
+    def SolveSolutionStep(self):
+        ## Obtain scipy matrices
+        # M = self._MassMatrixComputation()
```
Comment (armingeiser) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: why commented?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    def _create_linear_solver(self):
+        ''' Linear solver will not be used. But eventually the solution stratgey calls the solver's clear function. 
+        To avoid crashing linear solver is provided here'''
+        return linear_solver_factory.CreateFastestAvailableDirectLinearSolver()
+
+    def _create_mechanical_solution_strategy(self):
+        if self.settings["block_builder"].GetBool() == True:
+            warn_msg = "In case an eigenvalue problem is computed an elimantion builder shall be used to ensure boundary conditions are applied correctly!"
+            KratosMultiphysics.Logger.PrintWarning("CustomScipySolver", warn_msg)
```
Comment (armingeiser) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: Would an error work in this case? Or do you need the possibility to use a 'block_builder'

**Review by armingeiser — COMMENTED** [review]
Looks good for me in general, i have only some small comments

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 4 above)
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: Because that function should be overridden in the derived class. I just wanted to show how it works. Shall i uncomment it?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 5 above)
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: Depends on what the user wants to do. I did it this way to be more general. In case user wants to have full matrix. If elimination builder only is enough error message would be better.

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 4 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: If this function should always be overridden and never called directly, raise an error here
```python
        raise Exception("This function has to be implemented in the derived class")
```
The way it could be implemented could be explained in a docstring for this function, using the `"""` see e.g. kratos/python.

File: kratos/python/add_strategies_to_python.cpp (hunk 1 above)
Comment (philbucher) [reply kratos/python/add_strategies_to_python.cpp]: yes please make a PR for the changes in the Core

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 5 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: Ok for me.

Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: I don't understand why one cannot use the block-builder here

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        return this_defaults
+
+    #### Private functions ####
+    def _create_solution_scheme(self):
+        """Create the scheme for the scipy solver.
+
+        The scheme determines the mass and stiffness matrices 
+        """
+        scheme_type = self.settings["scheme_type"].GetString()
+        if scheme_type == "dynamic":
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: static is missing

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 4 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: I would call it only if you have a dynamic scheme

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            err_msg =  "The requested scheme type \"" + scheme_type + "\" is not available!\n"
+            err_msg += "Available options are: \"dynamic\""
+            raise Exception(err_msg)
+
+        return solution_scheme
+
+    def _create_linear_solver(self):
+        ''' Linear solver will not be used. But eventually the solution stratgey calls the solver's clear function. 
+        To avoid crashing linear solver is provided here'''
+        return linear_solver_factory.CreateFastestAvailableDirectLinearSolver()
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: if you only need a dummy then this should return the baseclass

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+from KratosMultiphysics import python_linear_solver_factory as linear_solver_factory
+
+# Import scipy modules
+import KratosMultiphysics.scipy_conversion_tools
+
+def CreateSolver(main_model_part, custom_settings):
+    return CustomScipySolver(main_model_part, custom_settings)
+
+class CustomScipySolver(MechanicalSolver):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]:
```suggestion
class MatrixBuilder(MechanicalSolver):
```

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py [hunk 9]
```diff
@@ -42,9 +42,12 @@ def CreateSolverByParameters(model, solver_settings, parallelism):
         elif solver_type == "adjoint_static":
             solver_module_name = "structural_mechanics_adjoint_static_solver"
 
+        elif solver_type == "custom_scipy":
+            solver_module_name = "structural_mechanics_custom_scipy_solver"
+
         else:
             err_msg =  "The requested solver type \"" + solver_type + "\" is not in the python solvers wrapper\n"
-            err_msg += "Available options are: \"static\", \"dynamic\", \"eigen_value\", \"harmonic_analysis\", \"formfinding\", \"adjoint_static\""
+            err_msg += "Available options are: \"static\", \"custom_scipy\", \"dynamic\", \"eigen_value\", \"harmonic_analysis\", \"formfinding\", \"adjoint_static\""
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: I don't think it should be here, 
Reason: this is quite sth special, that cannot be used without a corresponding `AnalysisStage`
Which btw I think you should add in this PR

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 8 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: :+1: or ScipyMatrixBuilder

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 5 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: it is not strictly right that you cannot use it. the problem is that the eigenvalue associated to the diagonal which corresponds to the dirichlet modes(the "1s" in the diagonal is only controlled by the  value of the number you set in there. Depending on the chosen value it may interfere with the other eigenvalues

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py (hunk 9 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: why do you need a special analysis stage? I do think that one should be able to construct it...that would be very good for experiencing

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 8 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: Ouch i tend to disagree on this. this is a complete solver, which can be integrated together with the standard analysis stage...

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 4 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: this requires a special scheme...which is designed on purpose to compute the mass matrix and the stiffness matrix

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 7 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: i agree

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py (hunk 9 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: my point is how do you use this solver?
It extends the interface of `PythonSolver` with the access-fct to the mass & stiffness matrix
=> this means that you need a custom analysis stage in any case, otherwise it does not make sense using this solver, bcs except building the matrices it does (and should not) do anything

**Comment by manuelmessmer** [comment]
> why do you compute this by hand?
> we have this functionality already in StructuralMechanics

Do you mean the assignment of the eigenvectors?

**Comment by philbucher** [comment]
no the entire computation of the eigenvalues.

**Comment by philbucher** [comment]
https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py

**Comment by philbucher** [comment]
@rishithellathmeethal could you please take a look at this? It is basically the same what we did in a hackish way some weeks ago

**Comment by rishithellathmeethal** [comment]
> @rishithellathmeethal could you please take a look at this? It is basically the same what we did in a hackish way some weeks ago

ya, Its similar except that we wrote all these inside solvesolutionstep. 
I would like to have a function to get System vector too. Now I am accessing it from strategy using the "GetSystemVector"().

**Comment by manuelmessmer** [comment]
> https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py

Yes I know, but @RiccardoRossi wanted to solve the problem using scipy

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 8 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: Now, with the implementation on SolveSolutionStep it would be something like "ScipyEigenvalueSolver"

**Review by rishithellathmeethal — DISMISSED** [review]

**Comment by philbucher** [comment]
Ok so given the discussion for me the usage/goal is still a bit unclear (@RiccardoRossi maybe you could answer since I think you came up with the design idea)
Is this solver supposed to be used for 
    a) computing the stiffness (& mass) -matrix or 
    b) for solving the eigenvalue problem with scipy?
I think a) by itself is already very useful and would help many ppl. This howerver is a special usecase that cannot be integrated in the regular workflow, bcs it needs a custom `AnalysisStage` for calling the matrix-access-fcts

b) Adding b) I think is also fine even though I don't see the usecase as clear. I guess using scipy instead of our own eigenvalue-solver would be nice to have. (implementational detail: instead of reimplementing `_AssignVariables` the C++-version already existing in the `EigenSolverStrategy` should be used, which might need some refactor)

I think what could make sense is the following desing:
~~~py
class ScipyMatrixBuilder(MechanialSolver):
    def ComputeStiffnessMatrix(self):
        # compute stiffness matrix, to be called from a custom analysis_stage
        return stiffness_matrix

    def ComputeMassMatrix(self):
        # compute mass matrix, to be called from a custom analysis_stage
        return mass_matrix
  
class ScipyEigenSolver(ScipyMatrixBuilder):
    def SolveSolutionStep(self):
        K = self.ComputeStiffnessMatrix()
        M = self.ComputeMassMatrix()
        # solve eigenvalue problem with scipy
~~~

what do you think?

**Comment by RiccardoRossi** [comment]
Hi @philbucher i designed this as a "base" solver.
my idea is that someone can hack this soler by **creating a derived one** in which one can prototype, in python, what he wants.

the first usecase could be to compute eigenvalues, but the scope is much larger than that.

Essentially i am giving access to (a copy of) K,M (and as a second step i would like to add D) as well as to the doflist. 
one can do what he likes with those (in his own derived solver) and mock up in python a complete solver.

this is a very useful think to protype stuff at the python level, and definitely to teach for example about the application of constraints.

I am not in favor of splitting it as you propose, since the way it is done now the initialization and all the rest is done there where it should be.

**Comment by philbucher** [comment]
@RiccardoRossi ok this makes sense to me
Only thing is that IMO the computation of Eigenvalues should not be done in a baseclass, otherwise I agree with you

BTW this implies that we should finally make the usage of custom solvers easier => I have a solution for the same problem in the CoSimApp, I will try in the next days to port it also to Structural

**Comment by philbucher** [comment]
So I just talked to @RiccardoRossi and want to share what we discussed:
- the `CustomScipySolver` is to be considered as a baseclass for solvers with which "hacking" can be done in python. Hence it could be clearer to rename it to `CustomScipyBaseSolver` or sth similar
- the computation of the eigenvalues is an EXAMPLE to show a usecase (@manuelmessmer please leave then the AssignVariables` in python and put some comments making it clear that this is only an example for the usage
- I will implement the feature to construct custom solvers, then we don't need to add every solver by hand into the wrapper. I have this already implemented in CoSim, I only have to port this functionality

What I think is missing now is cleanup and I suggest also a test (using the eigenvalue computation)

**Comment by philbucher** [comment]
> I will implement the feature to construct custom solvers, then we don't need to add every solver by hand into the wrapper. I have this already implemented in CoSim, I only have to port this functionality

see #6250

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 8 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: true, nevertheless what is implemented there is just an example of what can be done... @manuelmessmer can u add a comment to tell that eigenvalue solution is only an example and that one should derive his own implementation from this "base solver"?

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py (hunk 9 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: as we spoke, this opens the door to experimenting with scipy. one should derive and use his experimental solver

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py (hunk 8 above)
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py]: Yes will do. I am also working on a test.
So I will keep the name "CustomScipySolver"?

File: applications/StructuralMechanicsApplication/tests/test_custom_scipy_base_solver.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    def test_eigen_with_constraints(self):
+        analysis_parameters_scipy = KratosMultiphysics.Parameters("""{
+            "problem_data"    : {
+                "parallel_type" : "OpenMP",
+                "echo_level"    : 1,
+                "start_time"    : 0.0,
+                "end_time"      : 1.0
+            },
+            "solver_settings" : {
+                "solver_type"              : "structural_mechanics_custom_scipy_base_solver",
```
Comment (manuelmessmer) [inline applications/StructuralMechanicsApplication/tests/test_custom_scipy_base_solver.py]: Please note: Test will fail. As I just provide the name of the solver to suit #6250.

**Comment by RiccardoRossi** [comment]
Hi @philbucher 
Are we waiting for something to merge this?

**Comment by philbucher** [comment]
Your review on #6250

**Comment by philbucher** [comment]
@manuelmessmer since #6250 is merged can you please merge master here such that we can go ahead with this PR?

**Comment by manuelmessmer** [comment]
Do we need anything more here?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+        return solution_scheme
+
+    def _create_linear_solver(self):
+        ''' Linear solver will not be used. But eventually the solution strategy calls the solver's clear function. 
+        To avoid crashing linear solver is provided here'''
+        return KratosMultiphysics.LinearSolver()
+
+    def _create_mechanical_solution_strategy(self):
+        if self.settings["block_builder"].GetBool() == True:
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py]:
```suggestion
        if self.settings["block_builder"].GetBool():
```

more pythonic

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    Derived class must override the function SolveSolutionStep. In there the Mass and Stiffness matrices
+    can be obtained as scipy matrices.
+    The computation of the eigenvalue problem in this implementation is only an example how this solver is to be used.
+
+    See structural_mechanics_solver.py for more information.
+    """
+    def __init__(self, main_model_part, custom_settings):
+        # Construct the base solver.
+        super(CustomScipyBaseSolver, self).__init__(main_model_part, custom_settings)
+        self.space = KratosMultiphysics.UblasSparseSpace()
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py]: Any particular reason this is a member?
Unless you have one, I would remove it and use it locally
It is the baseclass, derived classes might not need it

File: applications/StructuralMechanicsApplication/tests/test_custom_scipy_base_solver.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    props[KratosMultiphysics.YOUNG_MODULUS] = 210e9
+    props[KratosMultiphysics.DENSITY] = 7850
+    props[StructuralMechanicsApplication.CROSS_AREA] = 0.01
+    props[KratosMultiphysics.POISSON_RATIO] = 0.30
+    props[StructuralMechanicsApplication.TORSIONAL_INERTIA] = 0.00001
+    props[StructuralMechanicsApplication.I22] = 0.00002
+    props[StructuralMechanicsApplication.I33] = 0.00001
+    props[StructuralMechanicsApplication.USE_CONSISTENT_MASS_MATRIX] = True
+
+    for i_elem, connectivity in enumerate(element_connectivities):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_custom_scipy_base_solver.py:144]: +1 for using `enumerate`

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py [hunk 14]
```diff
@@ -306,6 +308,8 @@ def AssembleTestSuites():
     # Eigenvalues Postprocessing Process test
     smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TTestPostprocessEigenvaluesProcess]))
     smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TestEigenSolverWithConstraints]))
+    # Custom Scipy Solver test
+    nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TestCustomScipyBaseSolver]))
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py:312]: how long does this test take on your machine?

**Review by philbucher — COMMENTED** [review]
minor comment otherwise I am positive about this PR!

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py (hunk 12 above)
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py]: Well it is used in _MassMatrixComputation and _StiffnessMatrixComputation, so probably also the derived classes will need it, correct?
But I can also declare it as a local variable in both methods?

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py (hunk 14 above)
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py:312]: I believe to remember it was relatively long, although it is pretty much the same example as in the normal EigensolverTest.  I let you know, but first have to recompile.

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py (hunk 12 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py]: I would make them local variables
Note that the methods in the Space are all `static` hence I think this makes sense

Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py]: Ok will do!

**Comment by philbucher** [comment]
@manuelmessmer check codacy, it has some suggestions

Esp I would set the editor to auto delete whitespaces when saving a file. With VSCode this is very easy

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py (hunk 14 above)
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py:312]: It takes 0.5sec

Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py:312]: ok then nightly is appropriate

**Review by philbucher — APPROVED** [review]
Approving since this has been there for a long time and I think everyone had time to raise their opinion.
Nice work and thanks for the patience @manuelmessmer

**Comment by philbucher** [comment]
@manuelmessmer the test fails if scipy is not available
can you please add a `skiptest`?

**Comment by manuelmessmer** [comment]
I pushed the changes here #6791.
